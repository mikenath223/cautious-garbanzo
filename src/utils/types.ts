interface IRequest {
  type: string;
  query: string;
}

interface ILocation {
  name: string;
  country: string;
  region: string;
  localtime: string;
}

interface ICurrent {
  observation_time: string;
  temperature: number;
  weather_code: number;
  weather_icons: string[];
  weather_descriptions: string[];
  wind_speed: number;
  wind_degree: number;
  wind_dir: string;
  pressure: number;
  precip: number;
  humidity: number;
  cloudcover: number;
  feelslike: number;
  uv_index: number;
  visibility: number;
}

export interface IWeatherInfo {
  request: IRequest;
  location: ILocation;
  current: ICurrent;
}

export interface PromiseFulfilledResult<T> {
  status: "fulfilled";
  value: T;
}

interface PromiseRejectedResult {
  status: "rejected";
  reason: any;
}

export type PromiseSettledResult<T> = PromiseFulfilledResult<T> | PromiseRejectedResult;

export interface IResult {
  status: "fulfilled" | "rejected";
  value: IWeatherInfo
}

export interface IFavCitiesTemp {
  name: string | undefined,
  temperature: number | undefined,
}

export interface INote {
  id: string;
  text: string
}

export interface IPResponse {
  latitudeData: number;
  longitudeData: number;
  city: string;
  region: string;
  country_name: string;
  country_code: string;
}
