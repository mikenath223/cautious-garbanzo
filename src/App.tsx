import { BrowserRouter as Router, Routes, Route } from "react-router-dom";
import { routes as appRoutes } from "./routes";
import Navbar from "./components/Navbar";
import Footer from "./components/Footer";

export default function App() {
  return (
    <Router>
      <Navbar />
      <Routes>
        {appRoutes.map((route) => (
          <Route
            key={route.key}
            path={route.path}
            element={<route.component />}
          />
        ))}
      </Routes>
      <Footer />
    </Router>
  );
}
