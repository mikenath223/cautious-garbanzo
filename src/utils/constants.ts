import { IWeatherInfo } from "./types"

export const initFavCities: string[] = [
  "Tokyo",
  "Delhi",
  "Shanghai",
  "Dhaka",
  "Sao Paulo",
  "Mexico City",
  "Cairo",
  "Beijing",
  "Mumbai",
  "Osaka",
  "Chongqing",
  "Karachi",
  "Kinshasa",
  "Lagos",
  "Istanbul"
]

export const defaultCity: IWeatherInfo = {
  coord:
    { lon: 3.75, lat: 6.5833 },
  weather: [{ main: "Rain", description: "light rain", icon: "10n" }],
  main: {
    temp: 25.18, feels_like: 26.26, temp_min: 25.18, temp_max: 25.18, pressure: 1015, humidity: 96,
    sea_level: 1015, grnd_level: 1015
  },
  visibility: 10000,
  wind: { speed: 1.44, deg: 225, gust: 2.57 },
  sys: { country: "NG" },
  name: "Lagos"
}

export const defaultNote = {
  id: "01",
  text: "Welcome"
}
