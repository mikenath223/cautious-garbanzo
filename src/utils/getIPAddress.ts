const fetchIPAddress = async () => {
  try {
    const response = await fetch('https://ipapi.co/json')
    const data = await response.json()
    return data
  } catch (err) {
    console.error(err)
  }
}

export default fetchIPAddress
