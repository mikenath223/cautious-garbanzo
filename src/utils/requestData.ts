import { IWeatherInfo, PromiseFulfilledResult } from "./types"

export const getCityWeather = async (city: string, isCheckCache: boolean): Promise<IWeatherInfo> => {
  try {
    const currentDate = new Date()
    const localStorageKey = `${city.toLowerCase().trim()}-${currentDate.getHours()}-${currentDate.getDate()}-${currentDate.getMonth()}-${currentDate.getFullYear()}`
    const cachedCityData = localStorage.getItem(localStorageKey)
    if (cachedCityData && isCheckCache) {
      const weatherData = JSON.parse(cachedCityData)
      return weatherData
    }
    const url = `http://api.weatherstack.com/current?access_key=04c73e2c81cd37fd215de627927cdd44&query=${city}`;
    const response = await fetch(url)
    const result = await response.json()
    if (result.error) {
      return result?.error?.info
    }
    localStorage.setItem(localStorageKey, JSON.stringify(result))
    return result
  } catch (error) {
    const errMsg = `Failed to retrieve ${city.trim()} data!`
    throw new Error(errMsg)
  }
}

const sortCities = (cities: string[]) => cities.sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()))

export const getFavsCityWeather = async (cities: string[]) => {
  const fetchRequests = sortCities(cities).map(city => getCityWeather(city, true));

  try {
    const citiesInfo = await Promise.allSettled(fetchRequests)
    const data = citiesInfo.filter(data => data.status === "fulfilled").map(data => (data as PromiseFulfilledResult<IWeatherInfo>).value)
    return data
  } catch (_) {
    return []
  }
}