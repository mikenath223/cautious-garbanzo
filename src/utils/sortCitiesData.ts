import { IWeatherInfo } from "./types";

const sortCitiesData = (cities: IWeatherInfo[] | undefined) => cities?.sort((a, b) => a.name.toLowerCase().localeCompare(b.name.toLowerCase()))

export default sortCitiesData
