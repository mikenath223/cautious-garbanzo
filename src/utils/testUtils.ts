import { render } from '@testing-library/react';

export const withRender = (component: JSX.Element) =>
  render(component);

export const runMock = () => {
  const mockSuccessResponse = {};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    ok: true,
    json: () => mockJsonPromise,
  });

  global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
  jest.spyOn(localStorage, "setItem");
  localStorage.setItem = (key: string, value: string) => { };
  localStorage.getItem = (key: string) => null;
}