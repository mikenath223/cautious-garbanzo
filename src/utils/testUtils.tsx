import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

export const withRender = (component: JSX.Element) =>
  render(<MemoryRouter>{component}</MemoryRouter>);

export const runMock = () => {
  const mockSuccessResponse = {};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    ok: true,
    json: () => mockJsonPromise,
  });

  global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
  jest.spyOn(localStorage, "setItem");
  localStorage.setItem = (key: string, value: string) => {};
  localStorage.getItem = (key: string) => null;
};
