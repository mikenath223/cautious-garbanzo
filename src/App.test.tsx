import App from "./App";
import { withRender } from "./utils/testUtils";

describe("App", function () {
  it("should match snapshot", () => {
    const { container } = withRender(<App />);

    expect(container).toMatchSnapshot();
  });
});

// const mockFetchPromise = Promise.resolve({
//   json: () => Promise.resolve({}),
// });
// global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;

// jest.mock('lodash-es', () => ({
//   ...jest.requireActual('lodash-es'),
//   isEmpty: () => jest.fn(),
//   isNil: () => jest.fn()
// }));
