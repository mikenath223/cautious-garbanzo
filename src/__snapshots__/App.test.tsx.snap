// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div>
  <header
    class="header"
  >
    <nav
      class="nav max-width-inhibitor"
    >
      <h6
        class="app-name"
      >
        WeatherAPP
      </h6>
      <a
        aria-current="page"
        class="nav-link active"
        href="/"
      >
        <h6>
          Home
        </h6>
      </a>
    </nav>
  </header>
  <main
    class="max-width-inhibitor home-main"
  >
    <form
      class="search-form"
    >
      <input
        class="search-input"
        id="search"
        name="search"
        placeholder="Search city weather..."
        required=""
        type="text"
        value=""
      />
      <button
        class="submit-btn cursor-pointer"
        type="submit"
      >
        Search
      </button>
    </form>
    <section>
      <ul
        class="favCities"
      />
    </section>
  </main>
  <footer
    class="p-5"
  >
    <div
      class="footer max-width-inhibitor"
    >
      <h5>
        WeatherAPP 
        2023
         Â©
      </h5>
      <span
        class="italic font-medium"
      >
        Powered by weatherstack
      </span>
    </div>
  </footer>
</div>
`;
